### ISSUE TITLE
Metric reduction behaviour of `LightningModule.log()` across accelerator workers

### ISSUE THREAD

**Issue by low5545** [PyTorchLightning/pytorch-lightning#10122]
## 🐛 Bug

The metric reduction behaviour of `LightningModule.log()` is not too clearly documented. Thus, I ran the following 6 cases to gain a better understanding of it (applicable at least for `accelerator=ddp/ddp_spawn`):

Assuming `WORLD_SIZE=2` (tested on single node 2 GPUs)
#### Case 1
```python
def training/validation/test_step(self, batch, batch_idx):
    self.log("metric", self.global_rank, on_step=True, on_epoch=False, sync_dist=False)
    # TensorBoard logs:
    #     `metric(_step) = 0.0`, for all steps
```

#### Case 2
```python
def training/validation/test_step(self, batch, batch_idx):
    self.log("metric", self.global_rank, on_step=True, on_epoch=False, sync_dist=True)
    # TensorBoard logs:
    #     `metric(_step) = 0.5`, for all steps
```

#### Case 3
```python
def training/validation/test_step(self, batch, batch_idx):
    self.log("metric", self.global_rank, on_step=False, on_epoch=True, sync_dist=False)
    # TensorBoard logs:
    #     `metric(_epoch) = 0.5`, for all epochs
```

#### Case 4
```python
def training/validation/test_step(self, batch, batch_idx):
    self.log("metric", self.global_rank, on_step=False, on_epoch=True, sync_dist=True)
    # TensorBoard logs:
    #     `metric(_epoch) = 0.5`, for all epochs
```

#### Case 5
```python
def training/validation/test_step(self, batch, batch_idx):
    self.log("metric", self.global_rank, on_step=True, on_epoch=True, sync_dist=False)
    # TensorBoard logs:
    #     `metric_step = 0.0`, for all steps
    #     `metric_epoch = 0.5`, for all epochs
```

#### Case 6
```python
def training/validation/test_step(self, batch, batch_idx):
    self.log("metric", self.global_rank, on_step=True, on_epoch=True, sync_dist=True)
    # TensorBoard logs:
    #     `metric_step = 0.0`, for all steps
    #     `metric_epoch = 0.5`, for all epochs
```

#### Observations
- The training/validation/test step metric reduction behaviour in Case 1 & 2 aligns with the documentation
- From Case 3 & 4, it seems that epoch-accumulated metric reduction is done regardless of `sync_dist`. Hence, it also suggests that the `sync_dist` option only controls for the training/validation/test step metric reduction behaviour. This prompted me to think whether [this](https://pytorch-lightning.readthedocs.io/en/latest/advanced/multi_gpu.html#synchronize-validation-and-test-logging) is still relevant or not.
- With the observations from Case 1 to 4, I would expect Case 6 to give `metric_step = 0.5`, for all steps, but I was wrong. It now seems like by default, training/validation/test step metrics are not reduced, and epoch-accumulated metric are always reduced. Step metric reduction behaviour can only be changed via `sync_dist` when no epoch-accumulated metrics are being logged (i.e. `on_epoch=False`). Is this a bug?

### Environment

- PyTorch Lightning Version (e.g., 1.3.0): 1.4.5 / main
- PyTorch Version (e.g., 1.8) 1.9.0 / 1.10.0
- Python version: 3.8
- OS (e.g., Linux): Ubuntu 20.04
- CUDA/cuDNN version: 11.1.74 / 10.2
- GPU models and configuration: 2 x GTX1080Ti
- How you installed PyTorch (`conda`, `pip`, source): `conda`
- If compiling from source, the output of `torch.__config__.show()`:
- Any other relevant information:
